drivers/spi/spi_ll_stm32.c

```diff
@@ -0,0 +1,442 @@
+#include <spi_ll_stm32.h>
+
+#define CONFIG_CFG(cfg)						\
+((const struct spi_stm32_config * const)(cfg)->dev->config->config_info)
+
+#define CONFIG_DATA(cfg)					\
+((struct spi_stm32_data * const)(cfg)->dev->driver_data)
+
+#ifdef CONFIG_SPI_STM32_INTERRUPT
+static void spi_stm32_transmit(SPI_TypeDef *spi, struct spi_stm32_data *data);
+static void spi_stm32_receive(SPI_TypeDef *spi, struct spi_stm32_data *data);
+
+static void spi_stm32_isr(void *arg)
+{
+	struct device * const dev = (struct device *) arg;
+	const struct spi_stm32_config *cfg = dev->config->config_info;
+	struct spi_stm32_data *data = dev->driver_data;
+	SPI_TypeDef *spi = cfg->spi;
+
+	if (LL_SPI_IsActiveFlag_TXE(spi) && spi_context_tx_on(&data->ctx)) {
+		spi_stm32_transmit(spi, data);
+	}
+
+	if (LL_SPI_IsActiveFlag_RXNE(spi)) {
+		if (spi_context_rx_on(&data->ctx)) {
+			spi_stm32_receive(spi, data);
+		} else {
+			LL_SPI_DisableIT_RXNE(spi);
+		}
+	}
+
+	if (!spi_context_tx_on(&data->ctx) &&
+	    !spi_context_rx_on(&data->ctx)) {
+		LL_SPI_DisableIT_TXE(spi);
+		LL_SPI_DisableIT_RXNE(spi);
+
+		spi_context_cs_control(&data->ctx, false);
+
+		spi_context_complete(&data->ctx, 0);
+	}
+}
+#endif
+
+static int spi_stm32_configure(struct spi_config *config)
+{
+	const struct spi_stm32_config *cfg = CONFIG_CFG(config);
+	struct spi_stm32_data *data = CONFIG_DATA(config);
+	const u32_t scaler[] = {
+		LL_SPI_BAUDRATEPRESCALER_DIV2,
+		LL_SPI_BAUDRATEPRESCALER_DIV4,
+		LL_SPI_BAUDRATEPRESCALER_DIV8,
+		LL_SPI_BAUDRATEPRESCALER_DIV16,
+		LL_SPI_BAUDRATEPRESCALER_DIV32,
+		LL_SPI_BAUDRATEPRESCALER_DIV64,
+		LL_SPI_BAUDRATEPRESCALER_DIV128,
+		LL_SPI_BAUDRATEPRESCALER_DIV256
+	};
+	SPI_TypeDef *spi = cfg->spi;
+	u32_t clock;
+	int br;
+
+	if (spi_context_configured(&data->ctx, config)) {
+		/* Nothing to do */
+		return 0;
+	}
+
+	if (SPI_WORD_SIZE_GET(config->operation) != 8) {
+		return -ENOTSUP;
+	}
+
+	clock_control_get_rate(device_get_binding(STM32_CLOCK_CONTROL_NAME),
+			       (clock_control_subsys_t) &cfg->pclken, &clock);
+
+	for (br = 1 ; br <= ARRAY_SIZE(scaler) ; ++br) {
+		u32_t clk = clock >> br;
+
+		if (clk < config->frequency) {
+			break;
+		}
+	}
+
+	if (br > ARRAY_SIZE(scaler)) {
+		return -ENOTSUP;
+	}
+
+	LL_SPI_Disable(spi);
+	LL_SPI_SetBaudRatePrescaler(spi, scaler[br - 1]);
+
+	if (SPI_MODE_GET(config->operation) ==  SPI_MODE_CPOL) {
+		LL_SPI_SetClockPolarity(spi, LL_SPI_POLARITY_HIGH);
+	} else {
+		LL_SPI_SetClockPolarity(spi, LL_SPI_POLARITY_LOW);
+	}
+
+	if (SPI_MODE_GET(config->operation) == SPI_MODE_CPHA) {
+		LL_SPI_SetClockPhase(spi, LL_SPI_PHASE_2EDGE);
+	} else {
+		LL_SPI_SetClockPhase(spi, LL_SPI_PHASE_1EDGE);
+	}
+
+	LL_SPI_SetTransferDirection(spi, LL_SPI_FULL_DUPLEX);
+
+
+	if (config->operation & SPI_TRANSFER_LSB) {
+		LL_SPI_SetTransferBitOrder(spi, LL_SPI_LSB_FIRST);
+	} else {
+		LL_SPI_SetTransferBitOrder(spi, LL_SPI_MSB_FIRST);
+	}
+
+	LL_SPI_DisableCRC(spi);
+
+	if (config->operation & SPI_OP_MODE_SLAVE) {
```

Comment:
It works when slave transceive data is supplied as as fast as possible and request are managed in a secondary thread, actually the same functional code is used in production right now as Slave mode exclusively.

Runtime switch from master to slave should be supported, but we may need some fixup since we leave the controller ON between multiple slave transceive requests to avoid glitches.

Anyway, it will be a good point to have a strong slave api fully using the ASYNC capability.